File: app/src/Store/Types/Prediction.ts
export type KMeansInfo = {
  params: {
    n_clusters: number;
  };
  selected_center: number[];
  centers: [number, number][];
  hull: [number, number][];
};

export type SkylineInfo = {
  sense: ('max' | 'min')[];
  frontier: [number, number][];
};

export type DBScanInfo = {
  hull: [number, number][];
  params: {
    eps: number;
    min_samples: number;
  };
};

export type LinearRegressionInfo = {
  threshold: number;
  coeff: number[];
  intercept: number;
};

export type AlgorithmInfo = KMeansInfo | DBScanInfo | LinearRegressionInfo | SkylineInfo;

export type Algorithms = 'KMeans' | 'DBScan' | 'LR' | 'BNL';

export type Intents = 'Cluster' | 'Outlier' | 'NonOutlier' | 'LR:within' | 'LR:outside' | 'Skyline';

export type BasePrediction = {
  rank: number;
  intent: Intents;
  algorithm: Algorithms;
  description: string;
  dimensions: string[];
  info: AlgorithmInfo;
  hash: string;
};

export type Prediction = BasePrediction & {
  memberIds: string[];
  membership: {
    ipns: string[];
    isnp: string[];
    matches: string[];
  };
};

export type Predictions = Prediction[];

export function getBasePrediction(prediction: Prediction) {
  const pred: Partial<Prediction> = { ...prediction };

  return pred;
}

File: backend/inference_core/prediction.py
from typing import List

from backend.utils.hash import getUIDForString


class Prediction(object):
    def __init__(
        self,
        rank,
        intent,
        algorithm,
        description,
        memberIds,
        dimensions,
        info,
        membership,
        hash=None,
    ) -> None:
        self.rank: float = rank
        self.intent: str = intent
        self.algorithm: str = algorithm
        self.description: str = description
        self.memberIds: List[str] = memberIds
        self.dimensions: List[str] = dimensions
        self.info: dict = info
        self.membership: dict = membership
        self.hash: str = self.get_hash()
        if hash:
            self.hash = hash

    def serialize(self):
        return self.__dict__

    def get_hash(self):
        hashString = "_".join(sorted(self.memberIds))
        hashString = f"{hashString}_{self.algorithm}_{self.intent}"
        return getUIDForString(hashString)
